### PR TITLE
Add ProtoConcatenator class for Protobuf & JSON

### PR DESCRIPTION
**Pull request by gray-matter**
Allows reading a complex message from Garmadon and outputting it as a single-level object in Protobuf or JSON

### CODE REVIEW

File: readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    private static <MessageType> MessageType concatInner(Collection<Message> messages,
+                   Function<Collection<Descriptors.FieldDescriptor>, MessageType> messageBuilder,
+                   BiConsumer<Map.Entry<Descriptors.FieldDescriptor, Object>, MessageType> contentsConsumer) {
+        Collection<Map.Entry<Descriptors.FieldDescriptor, Object>> allFields = new HashSet<>();
+
+        for (Message message : messages) {
+            allFields.addAll(message.getAllFields().entrySet());
+        }
+
+        Collection<Descriptors.FieldDescriptor> allKeys = new LinkedList<>();
```
Comment (jpbempel) [inline readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: ArrayList?

Comment (gray-matter) [reply readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: I don't really mind either way, not sure it makes a big difference given I'm only doing additions and traversals

File: readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+public class ProtoConcatenator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProtoConcatenator.class);
+
+    /**
+     * Concatenate Protobuf messages into a single Protobuf message.
+     * /!\ Doesn't handle embedded objects and repeated fields /!\
```
Comment (ashangit) [inline readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: looks like that you manage it, no?

File: readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        }
+        catch (IllegalArgumentException e) {
+            LOGGER.error("Could not flatten Protobuf event", e);
+            return null;
+        }
+    }
+
+    /**
+     * Concatenate Protobuf messages into a single String -> object map.
+     * /!\ Doesn't handle embedded objects and repeated fields /!\
```
Comment (ashangit) [inline readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: idem

File: readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java (hunk 2 above)
Comment (gray-matter) [reply readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: I always underestimate my code :smile: Fixed

File: readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java (hunk 3 above)
Comment (gray-matter) [reply readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java]: idem

File: readers/common/pom.xml [hunk 4]
```diff
@@ -51,6 +51,13 @@
             <artifactId>garmadon-schema</artifactId>
             <version>${garmadon.version}</version>
         </dependency>
+
+        <!-- Protobuf -->
+        <dependency>
+            <groupId>com.github.os72</groupId>
+            <artifactId>protobuf-dynamic</artifactId>
+            <version>0.9.4</version>
```
[the remark this answers is not in the document]
Comment (gray-matter) [reply readers/common/pom.xml]: Fixed

**Pull request merged**